docs/wslact.1
.TH "WSLACT" "1" "DATEPLACEHOLDER" "VERSIONPLACEHOLDER" "WSL Utilities User Manual"
.SH NAME
.B wslact
- Component of WSL Utilities
.SH SYNOPSIS
.B wslact
.RB COMMAND
.I ...
.SH DESCRIPTION
A set of quick actions for WSL such as creating startup tasks and manually sync time between Windows and WSL.
.PP
Currently, we have two features available:
.in +4n
.nf
\fBts, time-sync\fR \- Time Sync
\fBam, auto-mount\fR \- Auto Mounting
\fBmr, memory-reclaim\fR \- Memory Reclamation
.fi
.in
.SH COMMANDS
.SS "Time Sync"
Time Sync (\fItime-sync\fR) feature allows user to manually sync time from Windows to WSL. Sometime, when you have network issues, WSL time will be slower than the Windows time. This tool is designed for this situation. Requires sudo.
.TP
SYNOPSIS
.B wslact time-sync
.RB [ \-h ]
.TP
OPTIONS
.nf
.B -h, --help
.in +4n
print a simple help.
.in
.fi
.SS "Auto Mounting"
Auto Mounting (\fIauto-mount\fR) feature allows you to mount all drives available on Windows to WSL. Requires sudo.
.TP
SYNOPSIS
.B wslact auto-mount
.RB [ \-mh ]
.TP
OPTIONS
.nf
.B -m, --mount-options
.in +4n
pass a list of options you want to pass to \fImount\fR command.
.in
.fi
.nf
.B -h, --help
.in +4n
print a simple help.
.in
.fi
.SS "Memory Reclamation"
Memory Reclamation (\fImemory-reclaim\fR) feature allows you to reclaim memory by dropping memory cache. Requires sudo.
.TP
SYNOPSIS
.B wslact memory-reclaim
.RB [ \-h ]
.TP
OPTIONS
.nf
.B -h, --help
.in +4n
print a simple help.
.in
.fi
.SH OPTIONS
.TP
.B -h, --help
print a simple help.
.TP
.B -v, --version
print current version.
.SH AUTHOR
Created by Patrick Wu <me@patrickwu.space>
.SH REPORTING BUGS
Report bugs to <https://github.com/wslutilities/wslu/issues>;
For Ubuntu specific/related bugs, report to <https://bugs.launchpad.net/ubuntu/+source/wslu>.
.SH COPYRIGHT
This is free software; you can redistribute it and/or modify it under
the terms of the GNU GPL version 3 or (at your option) any later
version.
There is NO warranty; not even MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.
.SH SEE ALSO
wslu(7)

debian/changelog
wslu (3.2.3-0kali1) kali-dev; urgency=medium

  * wslsys: fix an issue in call_windows_uptime
    (GH: #182)
  * header: fix a distro detection issue in 
    Ubuntu Impish (21.10)
  * debian/control: fix dependency issue in 
    installing wslu (LP: #1926771)
  * docs: include available commands in manpage 
    (GH: #169)
  * project: fix a problem that Powershell output
    non-unicode character (GH: #162)
  * wslview: fix an issue that caused the relative
    Linux path unable to be opened (GH: #165)
  * wslusc: fix an issue that some people unable
    to open their GUI application (GH: #174)
  * header: Avoid error message when the file doesn't
    exist (LP: 1911379)

 -- Patrick Wu <me@patrickwu.space>  Mon, 14 Jun 2021 22:14:49 +0800

wslu (3.2.2-0kali1) kali-dev; urgency=medium

  * docs: include available commands in manpage 
    (GH: #169)
  * project: fix a problem that Powershell output
    non-unicode character (GH: #162)
  * wslview: fix an issue that caused the relative
    Linux path unable to be opened (GH: #165)
  * wslusc: fix an issue that some people unable
    to open their GUI application (GH: #174)

 -- Patrick Wu <me@patrickwu.space>  Mon, 15 Mar 2021 10:52:33 +0800

wslu (3.2.1-0kali1) kali-dev; urgency=medium

  * header: improvements on the system drive detection. (closes #133)
  * tests: update the test system and tests. (closes #125)
  * wslact: add back-compatibility for the auto-mounting feature.
  * wslview: fix an issue when handling http/https
  * wslview: fix a bug that it won't open file:///mnt/c/... or /mnt/c/...
  * wslusc: fix an issue that wsl.exe execute with wrongly passed param (close #138)
  * wslusc: fix the problem that the previous version that GUI shortcut generated won't work.
  * wslsys: fix an issue that the WSL Version shown is wrong.
  * Includes a wslview.desktop (PR #143 and #146)
  * Update the script to fix badly generated shortcut issue (fixes #141)
  * Fix a DPI script related space issue (fixes #144)

 -- Patrick Wu <me@patrickwu.space>  Wed, 23 Dec 2020 22:33:07 +0800

wslu (3.1.1-0kali1) kali-dev; urgency=medium

  * project: add separate Fedora Remix support
  * header: fix `ubuntu.exe` detection issue
  * wslview: mistake in calling a variable (fixes #118 again)
  * wslusc: always use the absolute path (fixes #129)
  * wslusc: fixes a possible white space path issue
  * wslusc: When TMP is empty, try TEMP

 -- Patrick Wu <me@patrickwu.space>  Wed, 01 Jul 2020 22:22:53 +0800

wslu (3.1.0-0kali1) kali-dev; urgency=medium

  * header: use the proper color definition
  * wslfetch: new display for WSL with WSL version displayed
  * wslfetch: use the proper logo for Ubuntu (Reported by @flexiondotorg)
  * wslsys: add WSL version option
  * wslsys: fix an issue in DPI scaling getting wrong value due to the API change in Latest fast ring insider build (Reported by @crramirez)
  * wslact: adding `--mount-options` for auto-mounting feature
  * wslact: fix an i18n related issue in time-sync feature (fixes #113 by @ricardosantos9521)
  * wslgsu: removed temporarily
  * wslgsu: limit the Windows version it can use
  * wslview: properly detect all different protocols and path styles (fixes #118)
  * package: fix alpine dependency issue (fixes #117)

 -- Patrick Wu <me@patrickwu.space>  Wed, 13 May 2020 22:53:03 +0800

wslu (3.0.1-0kali1) kali-dev; urgency=medium

  * Initial release

 -- Patrick Wu <me@patrickwu.space>  Mon, 13 Apr 2020 20:41:40 +0800


src/wslact.sh
# shellcheck shell=bash
help_short="wslact COMMAND ..."

function time_reset {
	local help_short="wslact time-reset [-h]"

	while [ "$1" != "" ]; do
		case "$1" in
			-h|--help) help "wslact" "$help_short"; exit;;
			*) shift;;
		esac
	done

	if [ "$EUID" -ne 0 ]; then
		error_echo "\`wslact time-sync\` requires you to run as root. Aborted." 1
	fi

	echo "${info} Before Sync: $(date +"%d %b %Y %T %Z")"
	if date -s "$(winps_exec "Get-Date -UFormat \"%m/%d/%Y %T %Z\"" | tr -d "\r")" >/dev/null; then
		echo "${info} After Sync: $(date +"%d %b %Y %T %Z")"
		echo "${info} Manual Time Reset Complete."
	else
		error_echo "Time Sync failed." 1
	fi
}

function auto_mount {
	local help_short="wslact auto-mount [-mh]"
	mntpt_prefix="$(interop_prefix)"
	sysdrv_prefix="$(sysdrive_prefix)"

	mount_opt=""
	while [ "$1" != "" ]; do
		case "$1" in
			-m|--mount-options) shift; mount_opt="$1"; shift;;
			-h|--help) help "wslact" "$help_short"; exit;;
			*) shift;;
		esac
	done

	if [ "$EUID" -ne 0 ]; then
		error_echo "\`wslact auto-mount\` requires you to run as root. Aborted." 1
	fi

	#shellcheck disable=SC1003
	drive_list="$("$mntpt_prefix$sysdrv_prefix"/WINDOWS/system32/fsutil.exe fsinfo drives | tail -1 | tr '[:upper:]' '[:lower:]' | tr -d ':\\' | sed -e 's/drives //g' -e "s|$sysdrv_prefix ||g" -e 's|\r||g' -e 's| $||g' -e 's| |\n|g')"

	if [ -n "$mount_opt" ]; then
		echo "${info} Custom mount option detected: $mount_opt"
	elif [ -f /etc/wsl.conf ]; then
		tmp="$(grep ^options /etc/wsl.conf | sed -r -e 's|^options[ ]+=[ ]+||g' -e 's|^"||g' -e 's|"$||g')"
		if [ "$tmp" != "" ]; then
			echo "${info} Custom mount option detected: $tmp"
			mount_opt="$tmp"
			unset tmp
		fi
	fi

	mount_s=0
	mount_f=0
	mount_j=0
	
	for drive in $drive_list; do
		[[ -d "$mntpt_prefix$drive" ]] || mkdir -p "$mntpt_prefix$drive"
		if [[ -n $(find "$mntpt_prefix$drive" -maxdepth 0 -type d -empty 2>/dev/null) ]]; then
			echo "${info} Mounting Drive ${drive^} to $mntpt_prefix$drive..."
			if mount -t drvfs "${drive}:" "$mntpt_prefix$drive" -o "$mount_opt" 2>/dev/null; then
				echo "${info} Mounted Drive ${drive^} to $mntpt_prefix$drive."
				mount_s=$((mount_s + 1))
			else
				echo "${error} Failed to mount Drive ${drive^}. Skipped."
				mount_f=$((mount_f + 1))
			fi
		else
			echo "${warn} Already mounted Drive ${drive^} at $mntpt_prefix$drive. Skipped."
			mount_j=$((mount_j + 1))
		fi
	done
	echo "${info} Auto mounting completed. $mount_s drive(s) succeed. $mount_f drive(s) failed. $mount_j drive(s) skipped."
}

function memory_reclaim {
	local help_short="wslact memory-reclaim [-h]"

	while [ "$1" != "" ]; do
		case "$1" in
			-h|--help) help "wslact" "$help_short"; exit;;
			*) shift;;
		esac
	done

	if [ "$EUID" -ne 0 ]; then
		error_echo "\`wslact memory-reclaim\` requires you to run as root. Aborted." 1
	fi

	sync
	echo 1 > /proc/sys/vm/drop_caches
	echo "${info} Memory Reclaimed."
}

while [ "$1" != "" ]; do
	case "$1" in
		ts|time-sync|tr|time-reset) time_reset "$@"; exit;;
		am|auto-mount|sm|smart-mount) auto_mount "$@"; exit;;
		mr|memory-reclaim|mem-reclaim) memory_reclaim "$@"; exit;;
		-h|--help) help "$0" "$help_short"; exit;;
		-v|--version) version; exit;;
		*) error_echo "Invalid Input. Aborted." 22;;
	esac
done
